benchmarks/newton_solver_benchmark_set/spiegelman_et_al_2016/drucker_prager_compositions.cc

```diff
@@ -0,0 +1,711 @@
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1];
+        }
+      return volume_fractions;
+    }
+
+    template <int dim>
+    double
+    DruckerPragerCompositions<dim>::
+    compute_second_invariant(const SymmetricTensor<2,dim> strain_rate, const double min_strain_rate) const
+    {
+      const double edot_ii_strict = std::sqrt(strain_rate*strain_rate);
+      const double edot_ii = std::max(edot_ii_strict, min_strain_rate*min_strain_rate);
+      return edot_ii;
+    }
+
+    template <int dim>
+    double
+    DruckerPragerCompositions<dim>::
+    compute_viscosity(const double edot_ii,const double pressure,const int comp,const double prefactor,const bool regularize, const double min_visc, const double max_visc) const
+    {
+      double viscosity;
+      if (comp == 0)
+        {
+          const double strength = ( (dim==3)
+                                    ?
+                                    ( 6.0 * cohesion[comp] * cos_phi[comp] + 6.0 * std::max(pressure,0.0) * sin_phi[comp] )
+                                    / ( sqrt_3 * ( 3.0 + sin_phi[comp] ) )
+                                    :
+                                    cohesion[comp] * cos_phi[comp] + std::max(pressure,0.0) * sin_phi[comp] );
+
+          // Rescale the viscosity back onto the yield surface
+          if (strength != 0 && edot_ii != 0)
+            viscosity = strength / ( 2.0 * sqrt_half * edot_ii );
+          else
+            viscosity = ref_visc;
+          if (regularize == true)
+            viscosity = ref_visc*viscosity / (ref_visc + viscosity);
+        }
+      else
+        {
+          viscosity = prefactor;
+        }
+
+      return std::max(std::min(viscosity,max_visc),min_visc);
+    }
+
+    template <int dim>
+    void
+    DruckerPragerCompositions<dim>::
+    evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+             MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      //set up additional output for the derivatives
+      MaterialModelDerivatives<dim> *derivatives;
+      derivatives = out.template get_additional_output<MaterialModelDerivatives<dim> >();
+
+      for (unsigned int i=0; i < in.temperature.size(); ++i)
+        {
+          // const Point<dim> position = in.position[i];
+          const double temperature = in.temperature[i];
+          const double pressure = in.pressure[i];
+
+          // Averaging composition-field dependent properties
+          // Compositions
+          //This assert may be help when designing tests for this material model
+          AssertThrow(in.composition[i].size()+1 == n_fields,
+                      ExcMessage("Number of compositional fields + 1 not equal to number of fields given in input file."));
+
+          const std::vector<double> volume_fractions = compute_volume_fractions(in.composition[i]);
+          double density = 0.0;
```

Comment:
One could use the `EquationOfState` module for many of the quantities below, but that might be quite a bit of work.